packages/components/src/custom-select-control-v2/legacy-component/index.tsx

```diff
@@ -12,18 +12,41 @@ import CustomSelectItem from '../item';
 import type { LegacyCustomSelectProps } from '../types';
 import * as Styled from '../styles';
 
+function useDeprecatedProps( {
+	__experimentalShowSelectedHint,
+	...otherProps
+}: LegacyCustomSelectProps ) {
+	return {
+		showSelectedHint: __experimentalShowSelectedHint,
+		...otherProps,
+	};
+}
+
+// The removal of `__experimentalHint` in favour of `hint` doesn't happen in
+// the `useDeprecatedProps` hook in order not to break consumers that rely
+// on object identity (see https://github.com/WordPress/gutenberg/pull/63248/files/175907e16310fcd3ef13fcb2964e138a69134760#r1668842238)
+function applyOptionDeprecations( {
+	__experimentalHint,
+	...rest
+}: LegacyCustomSelectProps[ 'options' ][ number ] ) {
+	return {
+		hint: __experimentalHint,
+		...rest,
+	};
+}
```

Comment:
@mirka Relatively to https://github.com/WordPress/gutenberg/pull/63248#discussion_r1672213131, I applied your suggestion of adapting the `options` array at runtime where necessary